include/evaluators/blas_tree_evaluator.hpp

```diff
@@ -0,0 +1,302 @@
+  using value_type = typename Expression::value_type;
+  using cont_type = typename Evaluator<LHS, Device>::cont_type;
+  /* constexpr static bool supported = LHS::supported && RHS::supported; */
+
+  Evaluator<LHS, Device> l;
+  Evaluator<RHS, Device> r;
+
+  Evaluator(Expression &expr)
+      : l(Evaluator<LHS, Device>(expr.l)), r(Evaluator<RHS, Device>(expr.r)) {}
+  size_t getSize() const { return r.getSize(); }
+  cont_type data() { return l.data(); }
+
+  bool eval_subexpr_if_needed(cont_type *cont, Device &dev) {
+    l.eval_subexpr_if_needed(NULL, dev);
+    r.eval_subexpr_if_needed(NULL, dev);
+    return true;
+  }
+
+  value_type eval(size_t i) {
+    l.eval(i);
+    return r.eval(i);
+  }
+  value_type eval(cl::sycl::nd_item<1> nditem) {
+    return eval(nditem.get_global(0));
+  }
+};
+
+template <class LHS, class RHS, typename Device>
+struct Evaluator<AssignExpr<LHS, RHS>, Device> {
+  using Expression = AssignExpr<LHS, RHS>;
+  using value_type = typename Expression::value_type;
+  using cont_type = typename Evaluator<LHS, Device>::cont_type;
+  /* constexpr static bool supported = LHS::supported && RHS::supported; */
+
+  Evaluator<LHS, Device> l;
+  Evaluator<RHS, Device> r;
+
+  Evaluator(Expression &expr)
+      : l(Evaluator<LHS, Device>(expr.l)), r(Evaluator<RHS, Device>(expr.r)) {}
+  size_t getSize() const { return r.getSize(); }
+  cont_type data() { return l.data(); }
+
+  bool eval_subexpr_if_needed(cont_type *cont, Device &dev) {
+    cont_type data = l.data();
+    l.eval_subexpr_if_needed(NULL, dev);
+    r.eval_subexpr_if_needed(&data, dev);
+    return true;
+  }
+
+  value_type eval(size_t i) { return l.eval(i) = r.eval(i); }
+  value_type eval(cl::sycl::nd_item<1> nditem) {
+    return eval(nditem.get_global(0));
+  }
+};
+
+template <class LHS1, class LHS2, class RHS1, class RHS2, typename Device>
+struct Evaluator<DoubleAssignExpr<LHS1, LHS2, RHS1, RHS2>, Device> {
+  using Expression = DoubleAssignExpr<LHS1, LHS2, RHS1, RHS2>;
+  using value_type = typename Expression::value_type;
+  using cont_type = typename Evaluator<LHS1, Device>::cont_type;
+  /* constexpr static bool supported = LHS1::supported && LHS2::supported &&
+   * RHS1::supported && RHS2::supported; */
+
+  Evaluator<LHS1, Device> l1;
+  Evaluator<LHS2, Device> l2;
+  Evaluator<RHS1, Device> r1;
+  Evaluator<RHS2, Device> r2;
+
+  Evaluator(Expression &expr)
+      : l1(Evaluator<LHS1, Device>(expr.l1)),
+        l2(Evaluator<LHS2, Device>(expr.l2)),
+        r1(Evaluator<RHS1, Device>(expr.r1)),
+        r2(Evaluator<RHS2, Device>(expr.r2)) {}
+  size_t getSize() const { return r1.getSize(); }
+  cont_type data() { return l1.data(); }
+
+  bool eval_subexpr_if_needed(cont_type *cont, Device &dev) {
+    l1.eval_subexpr_if_needed(NULL, dev);
+    l2.eval_subexpr_if_needed(NULL, dev);
+    cont_type data1 = l1.data(), data2 = l2.data();
+    r1.eval_subexpr_if_needed(&data1, dev);
+    r2.eval_subexpr_if_needed(&data2, dev);
+    return true;
```

Comment:
This is assignment part, I have not fully fixed it yet, when it comes to reduction. I will discuss it tomorrow. The idea is you either, reduce with assignment or use the result elsewhere.